sympy/utilities/lambdify.py

```diff
@@ -903,6 +919,123 @@ def _module_present(modname, modlist):
             return True
     return False
 
+def _cse_homogeneous(exprs, **kwargs):
+    """
+    Same as ``cse`` but the reduced_exprs are expres in the same type as 'exprs'
+
+    Parameters
+    ==========
+
+    exprs : list of sympy expressions, or a single sympy expression
+        The expressions to reduce.
+    kwargs : The over argument of the ``cse`` function
+
+    Returns
+    =======
+
+    replacements : list of (Symbol, expression) pairs
+        All of the common subexpressions that were replaced. Subexpressions
+        earlier in this list might show up in subexpressions later in this
+        list.
+    reduced_exprs : list of sympy expressions
+        The reduced expressions with all of the replacements above.
+
+    Examples
+    ========
+
+    >>> from sympy.utilities.lambdify import _cse_homogeneous
+    >>> from sympy import Symbol, cos, Tuple
+    >>> x = Symbol("x")
+    >>> _cse_homogeneous("cos(x + 1)") # str
+    ([], 'cos(x + 1)')
+    >>> _cse_homogeneous(cos(x + 1)) # Basic
+    ([], cos(x + 1))
+    >>> _cse_homogeneous(1) # number
+    ([], 1)
+    >>> _cse_homogeneous(Tuple(1, x))
+    ([], (1, x))
+    >>> _cse_homogeneous([1, x])
+    ([], [1, x])
+    >>> _cse_homogeneous((1, x))
+    ([], (1, x))
+    >>> _cse_homogeneous({1, x})
+    ([], {1, x})
+    """
+    from sympy import cse, sympify
+
+    if isinstance(exprs, str):
+        replacements, reduced_exprs = _cse_homogeneous(sympify(exprs), **kwargs)
+        return replacements, repr(reduced_exprs)
+    if isinstance(exprs, (list, tuple, set)):
+        replacements, reduced_exprs = cse(exprs, **kwargs)
+        return replacements, type(exprs)(reduced_exprs)
+    if isinstance(exprs, dict):
+        keys = list(exprs.keys()) # In order to guarantee the order of the elements.
+        replacements, values = cse([exprs[k] for k in keys], **kwargs)
+        reduced_exprs = {k: v for k, v in zip(keys, values)}
+        return replacements, reduced_exprs
+
+    try:
+        replacements, (reduced_exprs,) = cse(exprs, **kwargs)
```

Comment:
Are there tests for lambdify taking non-SymPy arguments like str or mpf? I think it should use `_sympify` on `expr` so that things like mpf are converted to SymPy objects first (and it shouldn't accept str at all).